### PR TITLE
Fix Amnesia implementation

### PR DESCRIPTION
**Pull request by Screaper91**
- Fixes GA prize cards disabling user's attacks
- Fixes Amnesia disabling nothing when used by Ren/Cleo
- Fixes Amnesia blocking the same attack consecutively

### CODE REVIEW

File: common/models/game-model.ts [hunk 1]
```diff
@@ -11,6 +11,32 @@ import ComponentTable from '../types/ecs'
...
+	default: () => T
+	values: Record<string, T> = {}
+
+	public constructor(defaultFactory: () => T) {
+		this.default = defaultFactory
+	}
+
+	public set(game: GameModel, value: T) {
+		this.values[game.id] = value
+		game.task
```
Comment (Lunarmagpie) [inline common/models/game-model.ts:25]:
```suggestion

```

**Review by Lunarmagpie — APPROVED** [review]
Thank you! I like how this is implemented.
I would consider combining InstancedValue and GameValue into one class. The type used as the key could simply be a generic. I would call this  base class `DefaultDict` because this is what its called in Python. if you don't want to make this change, or feel like its not a part of your patch, I'm happy to mark and something to do in the future.

File: common/cards/alter-egos-iii/hermits/architectfalse-rare.ts [hunk 2]
```diff
@@ -37,36 +49,51 @@ class ArchitectFalseRare extends Card {
...
 
-	lastAttacker = new InstancedValue<CardComponent | null>(() => null)
-	lastAttackType = new InstancedValue<HermitAttackType | null>(() => null)
+	lastAttackInfo = new GameValue<Record<PlayerEntity, AttackInfo | undefined>>(() => {
+		return {}
+	})
 
 	override onCreate(game: GameModel, component: CardComponent) {
-		const {opponentPlayer} = component
+		if (Object.hasOwn(this.lastAttackInfo, game.id)) return
```
Comment (Lunarmagpie) [inline common/cards/alter-egos-iii/hermits/architectfalse-rare.ts:57]: Correct me if im wrong, but I believe this will always be false?

Comment (Screaper91) [reply common/cards/alter-egos-iii/hermits/architectfalse-rare.ts:57]: Most decks will have multiple copies of the same hermit, this should reduce allocating duplicate resources and prevent data loss if another GA CardComponent is created in the middle of the game

Comment (Lunarmagpie) [reply common/cards/alter-egos-iii/hermits/architectfalse-rare.ts:57]: I believe `Object.hasOwn` is checking the GameValue object and not the instance at the moment. i think for that intention this line needs to be `if (Object.hasOwn(this.lastAttackInfo.values, game.id))`.

**Comment by Lunarmagpie** [comment]
I just noticed that items are not removed from the `lastAttackInfo` dictionary. I think you will need to add a `onDestroy` method for cards when the component is destroyed, then clear the item from the `lastAttackInfo` dictionary there.

**Comment by Screaper91** [comment]
Currently components are never explicitly destroyed, only garbage collected when a game ends. That would require more work on the ecs system, but for now it will take a while for two games to generate the same `game.internalId`.

**Comment by Lunarmagpie** [comment]
Thats true. Although it probably won't cause an issue, I don't like leaving memory around forever, and I also like taking advantage of garbage collection when we can. I think the best way to fix this would be created an 'onGameEndHook` we can subscribe to.

**Comment by Lunarmagpie** [comment]
Thank you!

**Comment by Lunarmagpie** [comment]
I just tested this and it works! I am ready to merge this.

**Pull request merged**